File: 149. Max Points on a Line/slope.cpp
class Solution {
public:
    int maxPoints(vector<vector<int>>& arr) {
        int sol = INT_MIN;

        for (int i = 0; i < arr.size(); i++){
            unordered_map<double, int> mp;
            int temp_sol = 0;

            for (int j = 0; j < arr.size(); j++){

                if (i == j)
                    continue;

                double y_dis = arr[j][1] - arr[i][1];
                double x_dis = arr[j][0] - arr[i][0];

                if (x_dis == 0){
                    mp[INT_MAX]++;

                    temp_sol = max(temp_sol, mp[INT_MAX]);

                    continue;
                }

                double slope = y_dis / x_dis;
                mp[slope]++;
                temp_sol = max(temp_sol, mp[slope]);
            }

            sol = max(sol, temp_sol + 1);
        }

        return sol;
    }
};

File: 149. Max Points on a Line/slope.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        def find_unique_line(point_1, point_2):
            if point_2[0] - point_1[0] == 0:
                return point_1[0]

            slope = (point_2[1] - point_1[1]) / (point_2[0] - point_1[0])
            bias = point_1[1] - point_1[0] * slope

            return (slope, bias)

        result = 0

        for i in range(len(points)):

            count = defaultdict(int)

            for j in range(i + 1, len(points)):

                count[find_unique_line(points[i], points[j])] += 1

            if count:
                result = max(result, max(count.values()))

        return result + 1